uitools/cpp/Esri/ArcGISRuntime/Toolkit/SmartLocatorSearchSource.cpp

```diff
@@ -0,0 +1,202 @@
+  }
+
+  /*!
+    \brief Returns the minimum number of suggestions to attempt to retrieve. If there are too few results, the search is repeated with
+    loosened parameters until enough suggesttions are accumulated.
+
+    \note If no search is successful, it is still possible to have a total number of sugggestions less than this threshold.
+    Set to zero to disable search repeat behavior. Defaults to 6.
+   */
+  int SmartLocatorSearchSource::repeatSuggestResultThreshold() const
+  {
+    return m_repeatSearchSuggestThreshold;
+  }
+
+  /*!
+    \brief Sets the minimum number of suggestions to attempt to retrieve
+    to \a repeatSearchSuggestThreshold.
+   */
+  void SmartLocatorSearchSource::setRepeatSuggestResultThreshold(int repeatSearchSuggestThreshold)
+  {
+    if (repeatSearchSuggestThreshold == m_repeatSearchSuggestThreshold)
+      return;
+
+    m_repeatSearchSuggestThreshold = repeatSearchSuggestThreshold;
+    emit repeatSuggestResultThresholdChanged();
+  }
+
+  /*!
+    \reimp
+   */
+  void SmartLocatorSearchSource::search(const SuggestResult& suggestion, const Geometry area)
+  {
+    // If area is specified, we listen in to the returned results and check to see if the number of
+    // results meets our threshold. If the number of results do not meet our threshold we re-do
+    // the search with the area constraint removed.
+    if (!area.isEmpty() && repeatSearchResultThreshold() > 0)
+    {
+      auto connection = std::make_shared<QMetaObject::Connection>(QMetaObject::Connection{});
+      *connection = connect(locator(),
+                            &LocatorTask::geocodeCompleted,
+                            this,
+                            [this, suggestion, area, connection](QUuid taskId, const QList<Esri::ArcGISRuntime::GeocodeResult>& geocodeResults)
+                            {
+                              // Warning: hack here. We can not detect in an async-friendly manner when a task is cancelled.
+                              // Which means we have to disconnect on "any" callback, otherwise we will build up space leaks with repeated
+                              // search-and-cancel behaviour.
+                              // This is obviously not ideal if the locator is used for multiple concurrent searches, but this is
+                              // an unlikely scenario.
+                              QObject::disconnect(*connection);
+
+                              if (searchTask().taskId() != taskId)
+                                return;
+
+                              if (geocodeResults.length() < repeatSearchResultThreshold())
+                              {
+                                auto params = geocodeParameters();
+                                params.setMaxResults(params.maxResults() - geocodeResults.length());
+                                searchTask() = locator()->geocodeWithSuggestResultAndParameters(suggestion, params);
```

Comment:
Should you just provide direct access to the member? This looks a bit strange to a first time reader. Presumably this function returns a reference to a member.